Time varying intercept

ulfaslak opened the pull request:
<!-- !! Thank your for opening a PR !! -->

<!--- Provide a self-contained summary of your changes in the Title above -->
<!--- This is what will be shown in the automatic release notes: https://github.com/pymc-labs/pymc-marketing/releases -->

This PR adds a time varying coefficient for the intercept. It's near-identical to #598, except it does not add support for time varying media contributions (this, we will add later).

## Description
The API is now:

```python
mmm = DelayedSaturatedMMM(
    date_column=my_date_column_name,
    channel_columns=my_media_column_names,
    time_varying_intercept=True,            # <-- This is new
)
```

Moreover, the `model_config` can now contain arguments for the time varying prior:

```Python
mmm = DelayedSaturatedMMM(
    ...
    model_config={
        ...,
        "tvp_kwargs": {
            "m": 200,
            "L": 100,  # If left to None, is estimated from the data (supporting 1 year of future OOS prediction)  
            "eta_lam": 1,
            "ls_mu": 52,  # If left to None, is estimated from the data (supporting 1 year of future OOS prediction)    
            "ls_sigma": 10,
            "cov_func": myCovFunc,  # If left to None, default Matern52 is used (good default).
        },
    }
)
```

Changes are also made to the plotting and data setter functions, to support out of sample predictions.

## Checklist
<!--- Make sure you have completed the following steps before submitting your PR -->
<!--- Feel free to type an `x` in all the boxes below to let us know you have completed the steps: -->
- [X] Checked that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html)
- [ ] Included tests that prove the fix is effective or that the new feature works
- [X] Added necessary documentation (docstrings and/or example notebooks)
- [ ] If you are a pro: each commit corresponds to a [relevant logical change](https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes)
<!--- You may find this guide helpful: https://mainmatter.com/blog/2021/05/26/keeping-a-clean-git-history/ -->

## Modules affected
<!--- Please list the modules that are affected by this PR by typing an `x` in the boxes below: -->
- [X] MMM
- [ ] CLV
<!--- Additionally, if you are a maintainer or reviewer, please make sure that the appropriate labels are added to this PR -->

## Type of change
<!--- Select one of the categories below by typing an `x` in the box -->
- [X] New feature / enhancement
- [ ] Bug fix
- [ ] Documentation
- [ ] Maintenance
- [ ] Other (please specify):
<!--- Additionally, if you are a maintainer or reviewer, please make sure that the appropriate labels are added to this PR -->


<!-- readthedocs-preview pymc-marketing start -->
----
📚 Documentation preview 📚: https://pymc-marketing--628.org.readthedocs.build/en/628/

<!-- readthedocs-preview pymc-marketing end -->

wd60622 commented on pymc_marketing/mmm/base.py:
```diff
@@ -54,13 +55,19 @@ def __init__(
+        self.X: Optional[pd.DataFrame] = None
+        self.y: Optional[Union[pd.Series, np.ndarray]] = None
+
+        self._time_resolution: int
+        self._time_index: NDArray[np.int_]
+        self._time_index_mid: int | float
```
I think that this would be nice to package up together. These three items are related and can get easily lost in the large class. 

It'd be a relatively small class but would lump some of these specific calculation together

Some pseudo code: 
```python
class TimeIndex: 
    def __init__(self) -> None: 
        self._time_resolution: int
        self._time_index: NDArray[np.int_]
        self._time_index_mid: int | float

    def initialize_values(X: pd.DataFrame, date_column: str) -> None:
        self._time_index = np.arange(0, X.shape[0])
        self._time_index_mid = X.shape[0] // 2
        self._time_resolution = (
            X[date_column].iloc[1] - X[date_column].iloc[0]
        ).days

    @property 
    def defaults(self): 
        L = (
            self._time_index_mid + DAYS_IN_YEAR / self._time_resolution
        )
        # more related calculation
        
    # More related methods
```

A single class instance could be an attribute of the MMM class then rather 3 different things that keep needing "used by TVP" etc

What are your thoughts?

juanitorduz replied:
I like it :)

ulfaslak replied:
Very nice. Will resolve all the little things, then revisit this.

ulfaslak replied:
Played around with an implementation of this, and will leave it for now. I think it makes more sense to add this in a later PR or as part of a major refactor.

wd60622 commented on pymc_marketing/mmm/delayed_saturated_mmm.py:
```diff
@@ -151,6 +178,11 @@ def _generate_and_preprocess_model_data(  # type: ignore
         }
         self.X: pd.DataFrame = X_data
         self.y: Union[pd.Series, np.ndarray] = y
+        self._time_index = np.arange(0, X.shape[0])
+        self._time_index_mid = X.shape[0] // 2
+        self._time_resolution = (
+            self.X[self.date_column].iloc[1] - self.X[self.date_column].iloc[0]
+        ).days
```
have this in a boolean block as well

wd60622 reviewed (commented):
Some comments to package up the code a bit more with extendability and reusability in mind in the future. Would love to hear the thoughts

juanitorduz commented on pymc_marketing/mmm/base.py:
```diff
@@ -343,19 +350,28 @@ def plot_posterior_predictive(
+                else self.get_target_transformer().transform(self.y[:, None]).flatten()  # type: ignore
             )
+
+            assert len(target_to_plot) == len(posterior_predictive_data.date), (
```
Can you use an if statement and raise an expectation instead? The assert statements can bring weird consequences, https://bandit.readthedocs.io/en/latest/plugins/b101_assert_used.html

juanitorduz replied:
can we wrap `self.get_target_transformer().transform(self.y[:, None]).flatten()` into a little function? (and add a test to document what we are doing?)

ulfaslak replied:
> Can you use an if statement and raise an expectation instead? The assert statements can bring weird consequences, https://bandit.readthedocs.io/en/latest/plugins/b101_assert_used.html
✅

> can we wrap self.get_target_transformer().transform(self.y[:, None]).flatten() into a little function? (and add a test to document what we are doing?)

I see, but not sure I agree. It's a simple expression (y a vertical vector, because that's what the transformer expects, then transform it into model scale and flatten it to an 1d array). @wd60622 thoughts?

wd60622 replied:
I am for making a small function. [This function](https://github.com/pymc-labs/pymc-marketing/blob/582d2ed4ddaa464cfe48dd52103afd5cb9001a4b/pymc_marketing/mmm/utils.py#L242) needs a friend 🤗
Something like `transform_1d_array`

ulfaslak replied:
Will make a util then :)

juanitorduz commented on pymc_marketing/mmm/delayed_saturated_mmm.py:
```diff
@@ -112,6 +121,24 @@ def _generate_and_preprocess_model_data(  # type: ignore
+            if `yearly_seasonality=True`.
+        y : Union[pd.Series, np.ndarray]
+            The target variable for the model (as provided).
+        _time_index : np.ndarray
+            The index of the date column. Used by TVP
+        _time_index_mid : int
+            The middle index of the date index. Used by TVP.
+        _time_resolution: int
+            The time resolution of the date index. Used by TVP.
```
I like the `class` idea by @wd60622 above

ulfaslak replied:
Gonna give it a shot 🚀

juanitorduz commented on pymc_marketing/mmm/delayed_saturated_mmm.py:
```diff
@@ -355,9 +387,41 @@ def build_model(
+                )
+
+                if self.model_config["tvp_kwargs"]["L"] is None:
+                    self.model_config["tvp_kwargs"]["L"] = (
```
Shall we use `c` instead of `L` as it has a relative interpretation range? https://www.pymc.io/projects/docs/en/stable/api/gp/generated/pymc.gp.HSGP.html

juanitorduz replied:
Also shall we use an `intercept_` prefix in the model config since we will have similar parameters when we add the media TVP ?

ulfaslak replied:
> Shall we use `c` instead of `L` as it has a relative interpretation range? https://www.pymc.io/projects/docs/en/stable/api/gp/generated/pymc.gp.HSGP.html

Since this *is* an MMM implementation, I think it's safe to set as an opinionated default that L reaches 1 year out of the sample (that's the default getting defined here). `c` is problematic, because it defines the basis function range as a function of the extent of your data. So if we e.g. set `c=2`, then with 1 year of training data, we extend the basis functions (and consequently how far we can theoretically predict into the future) a half year into the future. If we then have 2 years of training data, that extent becomes 1 year into the future. And so on. This is both computationally expensive, and not meaningful IMO.

I'd rather just say that, whatever the extent of your training data, we extend the basis functions 1 year into the future. No one are gonna want more, but if they do they can change L in the model config.

ulfaslak replied:
> Also shall we use an `intercept_` prefix in the model config since we will have similar parameters when we add the media TVP ?

So `self.model_config['intercept_tvp_kwargs`]`? Yes I thought about that to make space for `self.model_config['media_tvp_kwargs`]`, but my thought was, as this is an opinionated framework (at least I believe it should be), we can safely say that hyperparameters of time-varying priors are identical. The reason that I think this is a safe assumption is that media and intercept effect will probably vary over time with the same length scales? IDK, while writing this out now, I sense you have a point actually. Yeah we can change it.

juanitorduz replied:
- About using `L` instead of `c.` I like your arguments, let's leave `L` for now 👍 
- The statement "hyperparameters of time-varying priors are identical." might be to restrictive. I think `intercept_tvp_kwargs` give us more flexibility. I am open for discussion :) @wd60622 @carlosagostini ?

ulfaslak replied:
I see your point. It's definitely more forward compatible, and at any rate not something which will confuse the user (if anything it's self-documenting). I'll add it.

juanitorduz commented on pymc_marketing/mmm/delayed_saturated_mmm.py:
```diff
@@ -355,9 +387,41 @@ def build_model(
+                    self.model_config["tvp_kwargs"]["L"] = (
+                        self._time_index_mid + DAYS_IN_YEAR / self._time_resolution
+                    )
+                if self.model_config["tvp_kwargs"]["ls_mu"] is None:
```
same comment on the `intercept_`prefix

juanitorduz commented on pymc_marketing/mmm/delayed_saturated_mmm.py:
```diff
@@ -480,6 +549,14 @@ def default_model_config(self) -> Dict:
             "gamma_control": {"dist": "Normal", "kwargs": {"mu": 0, "sigma": 2}},
             "gamma_fourier": {"dist": "Laplace", "kwargs": {"mu": 0, "b": 1}},
+            "tvp_kwargs": {
+                "m": 200,
+                "L": None,
+                "eta_lam": 1,
+                "ls_mu": None,
+                "ls_sigma": 10,
+                "cov_func": None,
```
What do we expect to happen with these `None` defaults?

ulfaslak replied:
There's are checks starting on like 399, which set them if they are None. The reason that we wan't directly set them here is that the best defaults are estimated from the data, and we can't know here.

juanitorduz commented on pymc_marketing/mmm/tvp.py:
```diff
@@ -0,0 +1,67 @@
+    X_mid: int | float,
+    dims: Optional[tuple[str, str] | str] = None,
+    m: int = 40,
+    L: int | float = 100,
```
I suggest we use `c` instead

juanitorduz commented on pymc_marketing/mmm/tvp.py:
```diff
@@ -0,0 +1,67 @@
+
+    with pm.modelcontext(None) as model:
+        if cov_func is None:
+            eta = pm.Exponential(f"eta_{name}", lam=eta_lam)
+            ls = pm.InverseGamma(f"ls_{name}", mu=ls_mu, sigma=ls_sigma)
+            cov_func = eta**2 * pm.gp.cov.Matern52(1, ls=ls)
```
Can you mention these defaults on the docstrings 🙏

juanitorduz commented on pymc_marketing/mmm/tvp.py:
```diff
@@ -0,0 +1,67 @@
+    ls_sigma: float = 5,
+    cov_func: Optional[pm.gp.cov.Covariance] = None,
+) -> pm.Deterministic:
+    """Time varying prior, based the Hilbert Space Gaussian Process (HSGP).
```
Can you add a reference to the pymc classes and some relevant papers (you can take them from https://www.pymc.io/projects/docs/en/stable/api/gp/generated/pymc.gp.HSGP.html)

ulfaslak replied:
✅

juanitorduz commented on pymc_marketing/mmm/tvp.py:
```diff
@@ -0,0 +1,67 @@
+            hsgp_size = (n_columns, m)
+
+        gp = pm.gp.HSGP(m=[m], L=[L], cov_func=cov_func)
+        phi, sqrt_psd = gp.prior_linearized(Xs=X[:, None] - X_mid)
```
Will  `Xs=X[:, None] - X_mid` work for out-of-sample predictions?  Just a comment in view of https://github.com/pymc-devs/pymc/blob/main/pymc/gp/hsgp_approx.py#L288-L293 :)

ulfaslak replied:
Yes it works as long as `L` is set right (and I believe it currently is). Say you trained on two years of data, so `X` is the range `[0, 104]` (weeks) and `X_mid = 52`. Out of sample, `X_new` may be the range `[104 : 130]`, but that's all good as long as `L > max(X_new)` (and with basis functions extending 1 year into the future by default, it is).

juanitorduz replied:
All good! Let's use `L` then (see comment above)

juanitorduz commented on pymc_marketing/mmm/utils.py:
```diff
@@ -279,6 +281,10 @@ def apply_sklearn_transformer_across_dim(
     return data
 
 
+def softplus(x: pt.TensorVariable) -> pt.TensorVariable:
```
it seems to 🤔

ulfaslak replied:
you're right.

juanitorduz reviewed (commented):
@ulfaslak Super nice PR and fun to review! I left some small comments :)

juanitorduz commented:
@ulfaslak can you please the notebook name in https://raw.githubusercontent.com/pymc-labs/pymc-marketing/main/docs/source/notebooks/index.md so that it gets passed tot he docs 🙏

juanitorduz commented:
@ulfaslak you might also wanna update the branch as we solved an issue in https://github.com/pymc-labs/pymc-marketing/pull/633 🙏

ulfaslak commented on pymc_marketing/mmm/tvp.py:
```diff
@@ -0,0 +1,67 @@
+            cov_func = eta**2 * pm.gp.cov.Matern52(1, ls=ls)
+
+        hsgp_size: int | tuple[int, int] = m
+        if type(dims) is tuple:
```
✅

ulfaslak commented on pymc_marketing/mmm/delayed_saturated_mmm.py:
```diff
@@ -355,9 +387,41 @@ def build_model(
-                name="intercept", **self.model_config["intercept"]["kwargs"]
-            )
+            if self.time_varying_intercept:
+                time_index = pm.MutableData(
```
✅

ulfaslak commented:
> View / edit / reply to [this conversation](https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/628/discussion/#comment-2058785494) on ReviewNB
> 
> _juanitorduz commented on 2024-04-16T10:42:12Z_ ----------------------------------------------------------------
> 
> Line #7.        sampler_config={"chains": 2, "draws": 10, "tune": 10, "nuts_sampler": "numpyro"},
> Just 10 draws?

ya that was for debugging. I'm revising this whole notebook and making it nice now :). Thanks for all the comments!

ulfaslak commented:
Alright, final tests added to cover all new cases. @wd60622 @juanitorduz curious to hear your thoughts on these, maybe I'm breaking some implicit conventions for how we structure tests, please let me know and I will correct.

ulfaslak commented:
Ahh there are merge conflicts. Will resolve, but tests can still be reviewed

ulfaslak commented:
Oh, also forgot to add @wd60622 s cool idea for wrapping TVP params in a class. Will add ASAP, but after (and if tests are fine in your opinions) we are probably good to merge

cetagostini commented:
Just out of curiosity, have we tried to run the original notebook but now doing the intercept variable? I say this to know if we have seen the behaviour of the model in general with all parameters in place, controls, seasonality, channels, etc. 💡

wd60622 commented on pymc_marketing/mmm/delayed_saturated_mmm.py:
```diff
@@ -355,9 +390,41 @@ def build_model(
-            intercept = self.intercept_dist(
-                name="intercept", **self.model_config["intercept"]["kwargs"]
-            )
+            if self.time_varying_intercept:
```
Can we pull this all out into a function, say, `create_time_varying_intercept`. 
Think it'd be much easier to follow as a whole

```python
if self.time_varying_intercept: 
    intercept = create_time_varying_intercept(...)
    # or 
    intercept = pm.Deterministic(
         name="intercept", 
         var=create_time_varying_intercept(...), 
         dims="date",  
    )
else: 
    intercept = self.intercept_dist(
        name="intercept", 
        **self.model_config["intercept"]["kwargs"]
    )

ulfaslak replied:
This is neat, but has to be slightly less concise and elegant to accommodate for other time varying coefficients (like media channels).

ulfaslak replied:
✅

wd60622 reviewed (commented):
Lots of comments, sorry. 
This is gonna be epic 🚀 
Haven't touched the notebooks much but the plots are looking sweet

wd60622 commented on pymc_marketing/mmm/tvp.py:
```diff
@@ -0,0 +1,74 @@
+    ls_mu: float = 5,
+    ls_sigma: float = 5,
+    cov_func: pm.gp.cov.Covariance | None = None,
+) -> pm.Deterministic:
```
Deterministic is actually just a function: https://github.com/pymc-devs/pymc/blob/4bc84391893f1face230ed64241a339d4d9dbf62/pymc/model/core.py#L2126

ulfaslak replied:
Ah right. Seems right, will change it to TensorVariable.

ulfaslak commented on pymc_marketing/mmm/tvp.py:
```diff
@@ -0,0 +1,74 @@
+        Regression.
+    """
+
+    with pm.modelcontext(None) as model:
```
Don't fully understand. You mean:

```Python
model = pm.modelcontext(None)

if cov_func is None:
    eta = pm.Exponential(f"eta_{name}", lam=eta_lam)
    ls = pm.InverseGamma(f"ls_{name}", mu=ls_mu, sigma=ls_sigma)
    cov_func = eta**2 * pm.gp.cov.Matern52(1, ls=ls)
    
hsgp_size: int | tuple[int, int] = m
if isinstance(dims, tuple):
    n_columns = len(model.coords[dims[1]])
    hsgp_size = (n_columns, m)

gp = pm.gp.HSGP(m=[m], L=[L], cov_func=cov_func)
phi, sqrt_psd = gp.prior_linearized(Xs=X[:, None] - X_mid)
hsgp_coefs = pm.Normal(f"_hsgp_coefs_{name}", size=hsgp_size)
f = softplus(phi @ (hsgp_coefs * sqrt_psd).T)
return pm.Deterministic(name, f, dims=dims)
```

?

Or

```Python
model = pm.modelcontext(None)

if cov_func is None:
    eta = pm.Exponential(f"eta_{name}", lam=eta_lam)
    ls = pm.InverseGamma(f"ls_{name}", mu=ls_mu, sigma=ls_sigma)
    cov_func = eta**2 * pm.gp.cov.Matern52(1, ls=ls)

hsgp_size: int | tuple[int, int] = m
if isinstance(dims, tuple):
    n_columns = len(model.coords[dims[1]])
    hsgp_size = (n_columns, m)
  
with model:
    gp = pm.gp.HSGP(m=[m], L=[L], cov_func=cov_func)
    phi, sqrt_psd = gp.prior_linearized(Xs=X[:, None] - X_mid)
    hsgp_coefs = pm.Normal(f"_hsgp_coefs_{name}", size=hsgp_size)
    f = softplus(phi @ (hsgp_coefs * sqrt_psd).T)
    return pm.Deterministic(name, f, dims=dims)
```

Or something else?

wd60622 replied:
First one. Less indentation

ulfaslak commented on pymc_marketing/mmm/utils.py:
```diff
@@ -387,3 +407,13 @@ def create_new_spend_data(
     )
+
+
+def infer_time_index(
+    date_series_new: pd.Series, date_series: pd.Series, time_resolution: int
+) -> npt.NDArray[np.int_]:
+    """Infer the time-index given a new dataset.
+
+    Infers the time-indices by calculating the number of days since the first date in the dataset.
+    """
+    return (date_series_new - date_series[0]).dt.days.values // time_resolution
```
✅

ulfaslak commented:
> Lots of comments, sorry. This is gonna be epic 🚀 Haven't touched the notebooks much but the plots are looking sweet

Really nice comments man, thanks a lot. Addressed them all. Will add some more tests, give it a good stare and then it should be close to the finish line.

wd60622 commented on pymc_marketing/mmm/tvp.py:
```diff
@@ -0,0 +1,148 @@
+        model.add_coord("m", np.arange(m))  # type: ignore
+        hsgp_dims: str | tuple[str, str] = "m"
+        if isinstance(dims, tuple):
+            model.add_coord(dims[1], model.coords[dims[1]])
```
I was assuming this was already a dim in the model. Is that not right?
I feel that is fine assumption as it will error in not

wd60622 commented on pymc_marketing/mmm/delayed_saturated_mmm.py:
```diff
@@ -30,6 +31,8 @@
 
 __all__ = ["DelayedSaturatedMMM"]
 
+DAYS_IN_YEAR = 365.25
```
Duplicated now because of const file

wd60622 commented on pymc_marketing/mmm/tvp.py:
```diff
@@ -0,0 +1,148 @@
+        Model configuration.
+    """
+
+    with pm.modelcontext(None):
```
Some nitpick indentation comment as above. It can be assumed to be in model context because errors if more

wd60622 commented on pymc_marketing/mmm/tvp.py:
```diff
@@ -0,0 +1,148 @@
+        return pm.Deterministic(name, centered_f, dims=dims)
+
+
+def create_time_varying_intercept(
```
Test if isn't 🤗

ulfaslak commented:
pre-commit.ci autofix

ulfaslak commented:
@juanitorduz @wd60622 got the ✅ now. 

Updates are:
* Addressed all of @wd60622 s last comments. Good stuff 🙌.
* I updated the deterministic returned by `time_varying_prior` to be centered on one. This has some nice properties (see notebook).
* I added tests for new logic that wasn't covered.
* I updated the notebook to include an example where time varying intercept is a trend (not a sinus).

juanitorduz commented on pymc_marketing/mmm/tvp.py:
```diff
@@ -0,0 +1,148 @@
+    """
+
+    if X_mid is None:
+        X_mid = float(X.mean().eval())
```
This is related to https://github.com/pymc-devs/pymc/issues/7240 right/

wd60622 replied:
What aspect of it?

juanitorduz reviewed (commented):
The code part looks nice!

juanitorduz commented:
@ulfaslak This looks fantastic! The code itself is ✅ IMO (any comments @wd60622 ?). On the other hand, I think the notebook needs a bit more of the storyline. Consider a user who has never seen a time-varying intercept model.  The main points a reader should get from the notebook (IMO) are:

- What is a TV intercept?
- Why is it useful?
- How to use it in PyMC-Marketing?
-What are the benefits? (thinking about the MMM objective of better inference)
- When to use it?
- When it might not work?

WDYT?

PS. Also I added some minor comments about the plots XD

wd60622 commented on tests/mmm/test_tvp.py:
```diff
@@ -0,0 +1,123 @@
+
+        # Test that model can compile and sample
+        pm.Normal("obs", mu=prior, sigma=1, observed=np.random.randn(5, 3))
+        pm.sample(50, tune=50, chains=1)
```
```suggestion
        try: 
            pm.sample(50, tune=50, chains=1)
        except pm.SamplingError: 
            pytest.fail("Time varying parameter didn't sample")
```

wd60622 commented on tests/mmm/test_tvp.py:
```diff
@@ -0,0 +1,123 @@
+
+        # Test that model can compile and sample
+        pm.Normal("obs", mu=prior, sigma=1, observed=np.random.randn(5))
+        pm.sample(50, tune=50, chains=1)
```
```suggestion
        try: 
            pm.sample(50, tune=50, chains=1)
        except pm.SamplingError: 
            pytest.fail("Time varying parameter didn't sample")
```

wd60622 commented on pymc_marketing/mmm/tvp.py:
```diff
@@ -0,0 +1,148 @@
+
+    Returns
+    -------
+    pm.Deterministic
```
```suggestion
    pt.TensorVariable
```

ulfaslak replied:
nice catch

wd60622 commented on tests/mmm/test_tvp.py:
```diff
@@ -0,0 +1,123 @@
+        pm.sample(50, tune=50, chains=1)
+
+
+def test_calling_without_default_args(coords):
+    with pm.Model(coords=coords):
+        X = pm.Data("X", np.array([0, 1, 2, 3, 4]), dims="date")
+        time_varying_prior(name="test", X=X, dims="date")
```
Can this have some type of assert. Maybe the same as above

```suggestion
def test_calling_without_default_args(coords): 
    with pm.Model(coords=coords) as model: 
        X = pm.Data("X", np.array([0, 1, 2, 3, 4]), dims="date")
        prior = time_varying_prior(name="test", X=X, dims="date")
        
        # Assert output verification
        assert isinstance(prior, pt.TensorVariable)

        # Assert internal parameters are created correctly
        assert model.test_hsgp_coefs.eval().shape == (5,)

        # Assert default cov_func is used when none is provided
        assert "test_eta" in model.named_vars
        assert "test_ls" in model.named_vars
        assert "test_hsgp_coefs" in model.named_vars
```

wd60622 reviewed (commented):
Small suggestions
Taking a look at the notebooks now

wd60622 commented:
<p>+1</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/628/discussion/#comment-2076593385'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

ulfaslak commented:
> The main points a reader should get from the notebook (IMO) are:
> 
> * What is a TV intercept?
> * Why is it useful?
> * How to use it in PyMC-Marketing?
> * What are the benefits? (thinking about the MMM objective of better inference)
> * When to use it?
> * When it might not work?
> 
> Some of these questions have already been answer in the notebook. I think we just meet to complete the storyline and the takeaways 🤜🤛
> 
> WDYT?

Thank you. This is very actionable!

wd60622 commented:
The notebook looks really nice! @ulfaslak

ulfaslak commented:
<p>Isn't plt.show() more standard? I'll listen to a good argument (probably you're right).</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/628/discussion/#comment-2076593450'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>Well, it is not needed inside the notebook so that is the reason I suggest the change :)</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/628/discussion/#comment-2076593450'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

wd60622 commented:
Is there an example of changing the covariance function in the model config?

cetagostini commented:
<p>Agree!</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/628/discussion/#comment-2076889986'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

cetagostini commented:
<p>Agree with Juan here, I think it would be good to mention here what was discussed about the period within which the TVP does not deviate too much from reality. Being 90 days a great threshold if you work with daily data, of course not a consistent law for all cases, but regularly see it (Even in this example).</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/628/discussion/#comment-2076593576'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

cetagostini commented on pymc_marketing/mmm/delayed_saturated_mmm.py:
```diff
@@ -355,9 +388,23 @@ def build_model(
+                    "time_index",
+                    self._time_index,
+                    dims="date",
+                )
```
I believe `time_index` should be mutable no?

ulfaslak commented:
<p>Considering adding a third example with spiky events, cause really that's what this is good for, I think...</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/628/discussion/#comment-2076593559'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

ulfaslak commented:
<p>For sure. refactoring the whole thing now to avoid boilerplate</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/628/discussion/#comment-2076890035'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

ulfaslak commented:
<p>Love it!</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/628/discussion/#comment-2076889986'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

ulfaslak commented:
<p>Thanks</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/628/discussion/#comment-2076890054'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

ulfaslak commented:
<p>lazy/ultra-brief way to offset time on the x-axis</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/628/discussion/#comment-2076890105'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

ulfaslak commented:
<p>yea, we do write to data when we synth new targets, so another option is to write to data_train and data_test separately, but this seems like the lesser evil</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/628/discussion/#comment-2076890157'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

ulfaslak commented:
<p>resolved in latest commit, i believe</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/628/discussion/#comment-2081242984'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

ulfaslak commented:
@juanitorduz @wd60622 @carlosagostini Thanks for the reviews guys 🚀

I spent the day revising the notebook and I think it turned out quite nice:

* Updated story line. Much better read now, particularly intro IMO. Reflects most comments I believe, let me know if not.
* Everything factored into functions now that are reused across examples, to there is much less code to take in (fully DRY)
* Added an example on "random events" that I think you might appreciate. Plays with ls_mu and prior setting to show how to improve fit, with difficult intercept.

juanitorduz commented:
<p>+1 </p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/628/discussion/#comment-2076890089'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
<p>+1</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/628/discussion/#comment-2076890130'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
@ulfaslak the notebook is looking much better! I especially liked your last example! 💪 ! 

I left some minor comments. I think there are other small ones by Will. Once you check those (or ignore them if you think you are happy now and would rather work on iterations), we can merge this one :)

Also, can you add 

```python
%load_ext watermark
%watermark -n -u -v -iv -w -p pytensor
```

at the end as the other notebooks please :)

juanitorduz commented:
There is one test failing `tests/mmm/test_tvp.py::test_calling_without_default_args`

wd60622 commented:
<p>+1</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/628/discussion/#comment-2083477866'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

ulfaslak commented:
<p>done</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/628/discussion/#comment-2083477915'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

ulfaslak commented:
<p>nice catch</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/628/discussion/#comment-2087074487'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

ulfaslak commented:
> There is one test failing `tests/mmm/test_tvp.py::test_calling_without_default_args`

I'll address this tomorrow.

SO CLOSE!!

juanitorduz commented:
<p>That would be an interesting observation indeed!</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/628/discussion/#comment-2087074550'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
> > There is one test failing `tests/mmm/test_tvp.py::test_calling_without_default_args`
> 
> I'll address this tomorrow.
> 
> SO CLOSE!!

The notebook looks amazing! There are a couple of minor comments open in the notebook, but they are not blockers, IMO (it would be nice to test if, in this case, the out-of-sample forecast accuracy breaks after one length scale). So, once you are happy with the notebook and you fix the test, we can merge this amazing addition! :)

juanitorduz commented:
(tomorrow is 1st of May so maybe you do it on Thursday and enjoy the day :D )

cetagostini commented:
<p>I love the new introduction!</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/628/discussion/#comment-2087074487'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

ulfaslak commented:
<p>Maaa maybe. I think in reality things are more complicated than that and we can't make an observation that simple. The default length-scale is 2 years (104.36 when data is weekly, 730.5 when it is daily)... When I lower length-scale in the last example, to one year, OOS uncertainty actually get smaller. I could try to dig into the relationship between LS and OOS performance, but the finding and nuanced answer would probably be *it depends*, and then—to bake that into this notebook—would require a new section. That might clutter the narrative and would take me a bunch of time = slow down release. So while interesting that you note this, I am against adding it for now (could be material for another time though, let's take note).</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/628/discussion/#comment-2087074550'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

ulfaslak commented:
<p>I'll start using <code>ax.annotate</code> ALL THE TIME from now on 😈</p>

---

 <a href='https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/628/discussion/#comment-2087322487'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

juanitorduz commented:
> Maaa maybe. I think in reality things are more complicated than that and we can't make an observation that simple. The default length-scale is 2 years (104.36 when data is weekly, 730.5 when it is daily)... When I lower length-scale in the last example, to one year, OOS uncertainty actually get smaller. I could try to dig into the relationship between LS and OOS performance, but the finding and nuanced answer would probably be *it depends*, and then—to bake that into this notebook—would require a new section. That might clutter the narrative and would take me a bunch of time = slow down release. So while interesting that you note this, I am against adding it for now (could be material for another time though, let's take note).
> 
> [View entire conversation](https://app.reviewnb.com/pymc-labs/pymc-marketing/pull/628/discussion/#comment-2087074550) on ReviewNB

You are right! Let's skip this for this first iteration :)

ulfaslak commented:
> (tomorrow is 1st of May so maybe you do it on Thursday and enjoy the day :D )

Appreciate the gesture 🙏. I'm already two cups of coffee deep though, so I might as well get it wrapped up 😅

ulfaslak commented:
@juanitorduz @wd60622 @carlosagostini  OK I think it's done now. Fixed the test and did some last minute copy fixes using grammarly (lotta typos when writing in free hand in code editor). Letting the CI spinners converge now, and hopefully we can merge today on May 1st 🎉.

juanitorduz commented:
Fantastic 🚀 ! 

@wd60622 do you have any additional comments :) ?

wd60622 reviewed (approved):
Looks good! Epic

ulfaslak commented:
> > @juanitorduz @wd60622 @carlosagostini OK I think it's done now. Fixed the test and did some last minute copy fixes using grammarly (lotta typos when writing in free hand in code editor). Letting the CI spinners converge now, and hopefully we can merge today on May 1st 🎉.
> 
> How do you see Grammarly on notebooks? Or do you convert it to markdown?

I have Grammarly as a Chrome extension, and then I just load the notebook in the browser the old school way.
